experimental/calendar/src/iso.rs

```diff
@@ -0,0 +1,310 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{Calendar, Date, DateDuration, DurationUnit};
+use std::convert::{TryFrom, TryInto};
+
+#[derive(Copy, Clone, Debug)]
+/// The ISO Calendar
+pub struct Iso;
+
+/// A 1-indexed representation of an ISO day
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
+pub struct IsoDay(u8);
+/// A 1-indexed representation of an ISO month
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
+pub struct IsoMonth(u8);
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq, Ord, PartialOrd)]
+/// An ISO year. Year 0 == 1 BCE
+pub struct IsoYear(pub i32);
+
+impl TryFrom<u8> for IsoDay {
+    type Error = ();
+    fn try_from(int: u8) -> Result<Self, ()> {
+        if int > 31 || int < 1 {
+            return Err(());
+        }
+        Ok(Self(int))
+    }
+}
+
+impl TryFrom<u8> for IsoMonth {
+    type Error = ();
+    fn try_from(int: u8) -> Result<Self, ()> {
+        if int > 12 || int < 1 {
+            return Err(());
+        }
+        Ok(Self(int))
+    }
+}
+
+impl From<i32> for IsoYear {
+    fn from(int: i32) -> Self {
+        Self(int)
+    }
+}
+
+impl From<IsoDay> for u8 {
+    fn from(day: IsoDay) -> Self {
+        day.0
+    }
+}
+
+impl From<IsoMonth> for u8 {
+    fn from(month: IsoMonth) -> Self {
+        month.0
+    }
+}
+
+impl From<IsoYear> for i32 {
+    fn from(year: IsoYear) -> Self {
+        year.0
+    }
+}
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+pub struct IsoDateInner {
+    day: IsoDay,
+    month: IsoMonth,
+    year: IsoYear,
+}
+
+impl IsoDateInner {
+    fn add_months(&mut self, months: i32) {
+        // Get a zero-indexed new month
+        let new_month = (self.month.0 as i32 - 1) + months;
+        if new_month >= 0 {
+            self.year.0 += new_month / 12;
+            self.month.0 = ((new_month % 12) + 1) as u8;
+        } else {
+            // subtract full years
+            self.year.0 -= (-new_month) / 12;
+            // subtract a partial year
+            self.year.0 -= 1;
+            // adding 13 since months are 1-indexed
+            self.month.0 = (13 + (new_month % 12)) as u8
+        }
+    }
+}
+
+impl Calendar for Iso {
+    type DateInner = IsoDateInner;
+    fn date_from_iso(&self, iso: Date<Iso>) -> IsoDateInner {
```

Comment:
Suggestion (optional): make this function take `IsoDateInner` instead of `Date<Iso>` for consistency with the other functions that take DateInner?

Reply:
That doesn't work with the intent of this method: the intent is for _other calendars_ to be convertible to/from ISO. Code from calendar A should never have to care about the DateInner type for calendar B, since it'll be encapsulated. In the case of ISO-to-ISO it's a bit of a special (and mostly useless) case.